java/src/main/native/src/TableJni.cpp

```diff
@@ -189,6 +192,341 @@ typedef jni_table_writer_handle<cudf::io::detail::parquet::pq_chunked_state>
+    }
+    writer->WriteTable(*arrow_tab);
+  }
+
+  void close() {
+    if (initialized) {
+      writer->Close();
+      sink->Close();
+    }
+    initialized = false;
+  }
+};
+
+
+class jni_arrow_output_stream final : public arrow::io::OutputStream {
+public:
+  explicit jni_arrow_output_stream(JNIEnv *env, jobject callback) {
+    if (env->GetJavaVM(&jvm) < 0) {
+      throw std::runtime_error("GetJavaVM failed");
+    }
+
+    jclass cls = env->GetObjectClass(callback);
+    if (cls == nullptr) {
+      throw cudf::jni::jni_exception("class not found");
+    }
+
+    handle_buffer_method =
+        env->GetMethodID(cls, "handleBuffer", "(Lai/rapids/cudf/HostMemoryBuffer;J)V");
+    if (handle_buffer_method == nullptr) {
+      throw cudf::jni::jni_exception("handleBuffer method");
+    }
+
+    this->callback = env->NewGlobalRef(callback);
+    if (this->callback == nullptr) {
+      throw cudf::jni::jni_exception("global ref");
+    }
+  }
+
+  virtual ~jni_arrow_output_stream() {
+    // This should normally be called by a JVM thread. If the JVM environment is missing then this
+    // is likely being triggered by the C++ runtime during shutdown. In that case the JVM may
+    // already be destroyed and this thread should not try to attach to get an environment.
+    JNIEnv *env = nullptr;
+    if (jvm->GetEnv(reinterpret_cast<void **>(&env), cudf::jni::MINIMUM_JNI_VERSION) == JNI_OK) {
+      env->DeleteGlobalRef(callback);
+      if (current_buffer != nullptr) {
+        env->DeleteGlobalRef(current_buffer);
+      }
+    }
+    callback = nullptr;
+    current_buffer = nullptr;
+  }
+
+  arrow::Status Write(const std::shared_ptr<arrow::Buffer> & data) override {
+    return Write(data->data(), data->size());
+  }
+
+  arrow::Status Write(const void* data, int64_t nbytes) override {
+    JNIEnv *env = cudf::jni::get_jni_env(jvm);
+    int64_t left_to_copy = nbytes;
+    const char *copy_from = static_cast<const char *>(data);
+    while (left_to_copy > 0) {
+      long buffer_amount_available = current_buffer_len - current_buffer_written;
+      if (buffer_amount_available <= 0) {
+        // should never be < 0, but just to be safe
+        rotate_buffer(env);
+        buffer_amount_available = current_buffer_len - current_buffer_written;
+      }
+      long amount_to_copy =
+          left_to_copy < buffer_amount_available ? left_to_copy : buffer_amount_available;
+      char *copy_to = current_buffer_data + current_buffer_written;
+
+      std::memcpy(copy_to, copy_from, amount_to_copy);
+      copy_from = copy_from + amount_to_copy;
+      current_buffer_written += amount_to_copy;
+      total_written += amount_to_copy;
+      left_to_copy -= amount_to_copy;
+    }
+    return arrow::Status::OK();
+  }
+
+  arrow::Status Flush() override {
+    if (current_buffer_written > 0) {
+      JNIEnv *env = cudf::jni::get_jni_env(jvm);
+      handle_buffer(env, current_buffer, current_buffer_written);
+      current_buffer = nullptr;
```

Comment:
Don't we need to delete the global ref to the buffer before nulling the object pointer here?

Reply:
Probably.  I copied and pasted this code from `jni_writer_data_sink` so I will try to update it in both places.

Reply:
That caused errors with close called too many times.

Reply:
Never mind it was an issue with the AutoCloseable change